File: src/core/scheduler.ts
import { loadConfig, type Config } from "./config";
import { fmtTime } from "./format";
import { notify } from "./notify";
import { checkClaude, startClaude } from "./providers/claude";
import { checkCodex, startCodex } from "./providers/codex";
import { loadState, log, saveProviderState } from "./state";
import { PROVIDERS, type Provider, type ProviderState, type Snapshot } from "./types";

/** Don't send another start message within this long of the last one. */
const START_COOLDOWN_MS = 30 * 60_000;
const FAILURE_NOTIFY_AFTER = 3;
/**
 * Wait before confirming a start: right after a start, Codex's new window is
 * indistinguishable from its "no window" report (see IDLE_TOLERANCE_MS).
 */
export const timing = {
  confirmDelayMs: 90_000,
  /**
   * How long after a reported reset time to check again. Codex reports exact seconds;
   * Claude's /usage shows minutes only ("4:59pm" for a 17:00 reset), so it gets more
   * room: checking early would send the start message into the old session.
   */
  resetGraceMs: { claude: 90_000, codex: 30_000 } as Record<Provider, number>,
  /** Retry spacing when a session is still reported as running after its reset time. */
  resetRetryMs: 60_000,
};
const MAX_RESET_RETRIES = 5;

const BUSY = Symbol("busy");

export const LABEL: Record<Provider, string> = { claude: "Claude", codex: "Codex" };

export const check: Record<Provider, (cfg: Config) => Promise<Snapshot>> = { claude: checkClaude, codex: checkCodex };
export const start: Record<Provider, (cfg: Config) => Promise<string>> = { claude: startClaude, codex: startCodex };

function minutesOfDay(hhmm: string): number {
  const [h, m] = hhmm.split(":").map(Number);
  return h * 60 + (m || 0);
}

export function inActiveHours(cfg: Pick<Config, "activeHours">, now: number): boolean {
  if (!cfg.activeHours) return true;
  const d = new Date(now);
  const cur = d.getHours() * 60 + d.getMinutes();
  const s = minutesOfDay(cfg.activeHours.start);
  const e = minutesOfDay(cfg.activeHours.end);
  return s <= e ? cur >= s && cur < e : cur >= s || cur < e; // supports ranges past midnight
}

/** When the active-hours period containing `now` began, or null when `now` is outside it. */
export function activeHoursBegan(cfg: Pick<Config, "activeHours">, now: number): number | null {
  if (!cfg.activeHours || !inActiveHours(cfg, now)) return null;
  const s = minutesOfDay(cfg.activeHours.start);
  const d = new Date(now);
  d.setHours(Math.floor(s / 60), s % 60, 0, 0);
  if (d.getTime() > now) d.setDate(d.getDate() - 1); // began yesterday (range past midnight)
  return d.getTime();
}

export function shouldStart(
  snap: Snapshot,
  st: ProviderState,
  now: number,
  cfg: Pick<Config, "activeHours"> & Partial<Pick<Config, "autoStart">>,
): { start: boolean; reason: string } {
  if (snap.fiveHour.active) return { start: false, reason: "5h session active" };
  if (cfg.autoStart === false) return { start: false, reason: "auto-start paused" };
  if (snap.weekly && snap.weekly.usedPct >= 100) return { start: false, reason: "weekly limit exhausted" };
  if (!inActiveHours(cfg, now)) return { start: false, reason: "outside active hours" };
  if (st.lastStartAt && now - st.lastStartAt < START_COOLDOWN_MS) return { start: false, reason: "started recently" };
  return { start: true, reason: "5h session idle" };
}

/** Whether a snapshot shows a 5h session that's still running (unknown counts as not). */
export function sessionActive(snap: Snapshot | undefined, now: number): boolean {
  const w = snap?.fiveHour;
  if (!w || !w.active) return false;
  return w.resetsAt === null || w.resetsAt > now;
}

/**
 * Check one provider and, if its session is idle, start a new one. Mutates `st`.
 * `persist` is called as soon as a start succeeds, so the cooldown survives a quit or
 * crash during the confirmation delay.
 */
export async function tickProvider(
  p: Provider,
  cfg: Config,
  st: ProviderState,
  now = Date.now(),
  persist: (st: ProviderState) => void = () => {},
) {
  st.lastCheckAt = now;
  let phase: "check" | "start" = "check";
  try {
    const snap = await check[p](cfg);
    st.lastSnapshot = snap;
    const decision = shouldStart(snap, st, now, cfg);
    st.lastDecision = decision.reason;
    log({ provider: p, event: "check", fiveHour: snap.fiveHour, weekly: snap.weekly, decision: decision.reason });

    if (decision.start) {
      phase = "start";
      const reply = await start[p](cfg);
      st.lastStartAt = Date.now();
      persist(st);
      phase = "check";
      await Bun.sleep(timing.confirmDelayMs);
      const after = await check[p](cfg);
      st.lastSnapshot = after;
      st.lastDecision = after.fiveHour.active ? "5h session active" : "start sent, session not reported yet";
      log({ provider: p, event: "start", model: cfg[p].model, reply, fiveHour: after.fiveHour });
      notify(
        "Usage Window Starter",
        after.fiveHour.active
          ? `${LABEL[p]} 5h session started — resets ${fmtTime(after.fiveHour.resetsAt)}`
          : `${LABEL[p]} start message sent, but no active session reported yet`,
      );
    }
    st.consecutiveFailures = 0;
    st.failureNotified = false;
    st.lastError = undefined;
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    st.consecutiveFailures = (st.consecutiveFailures ?? 0) + 1;
    st.lastError = msg;
    log({ provider: p, event: "error", phase, error: msg, consecutive: st.consecutiveFailures });
    // A failed start (e.g. a model that's no longer offered) needs attention right away.
    if ((phase === "start" || st.consecutiveFailures >= FAILURE_NOTIFY_AFTER) && !st.failureNotified) {
      st.failureNotified = true;
      notify(
        "Usage Window Starter",
        phase === "start"
          ? `${LABEL[p]} couldn't start a 5h session with model "${cfg[p].model}": ${msg.slice(0, 100)}`
          : `${LABEL[p]} limit check failing: ${msg.slice(0, 120)}`,
      );
    }
  }
}

/**
 * Runs checks on the configured interval. Wakes every minute and compares wall-clock
 * time, so checks catch up right after the Mac wakes from sleep.
 */
export class Scheduler {
  private busy = new Set<Provider>();
  private timer: ReturnType<typeof setInterval> | null = null;
  private resetTimers = new Map<Provider, ReturnType<typeof setTimeout>>();
  private resetRetries = new Map<Provider, number>();

  constructor(private onChange: () => void = () => {}) {}

  run() {
    if (this.timer) return;
    this.timer = setInterval(() => void this.runDue(), 60_000);
    // The first runDue may skip providers checked recently (e.g. right after a restart),
    // so re-arm reset checks from the saved snapshots.
    const cfg = loadConfig();
    const state = loadState();
    for (const p of PROVIDERS) if (cfg[p].enabled) this.planResetCheck(p, state[p]?.lastSnapshot);
    void this.runDue();
  }

  stop() {
    if (this.timer) clearInterval(this.timer);
    this.timer = null;
    for (const t of this.resetTimers.values()) clearTimeout(t);
    this.resetTimers.clear();
  }

  isBusy(p: Provider) {
    return this.busy.has(p);
  }

  /**
   * Check every enabled provider whose interval has elapsed (or all of them with `force`).
   * Nothing runs outside active hours unless forced. When they begin, every provider not
   * checked since is due at once, instead of each waiting out its own interval.
   */
  async runDue(force = false, now = Date.now()) {
    const cfg = loadConfig();
    if (!force && !inActiveHours(cfg, now)) return;
    const began = activeHoursBegan(cfg, now);
    const state = loadState();
    const due = PROVIDERS.filter((p) => {
      if (!cfg[p].enabled || this.busy.has(p)) return false;
      const last = state[p]?.lastCheckAt ?? 0;
      return force || now - last >= cfg.intervalMin * 60_000 || (began !== null && last < began);
    });
    await Promise.all(
      due.map((p) => this.withProvider(p, (st) => tickProvider(p, cfg, st, now, (s) => saveProviderState(p, s)))),
    );
  }

  /** Start a session right now if it's idle, ignoring pause, cooldown and active hours. */
  async startNow(p: Provider) {
    const cfg = loadConfig();
    await this.withProvider(p, async (st) => {
      st.lastStartAt = undefined;
      st.failureNotified = false;
      await tickProvider(p, { ...cfg, autoStart: true, activeHours: null }, st, Date.now(), (s) =>
        saveProviderState(p, s),
      );
    });
  }

  /**
   * Send one test message with an unsaved config. It shares the provider lock with checks
   * and records a start, because a test message starts a session when none is running.
   */
  async testModel(p: Provider, cfg: Config): Promise<string> {
    const result = await this.withProvider(p, async (st) => {
      // Only a session that looked idle can have been started by the test. Recording it
      // during an active session would push the next auto-start back by the cooldown.
      const wasIdle = !sessionActive(st.lastSnapshot, Date.now());
      const reply = await start[p](cfg);
      if (wasIdle) st.lastStartAt = Date.now();
      log({ provider: p, event: "test", model: cfg[p].model, reply, countedAsStart: wasIdle });
      return reply;
    });
    if (result === BUSY) throw new Error(`${LABEL[p]} is busy with a check, try again in a moment`);
    return result;
  }

  /** Serialize work per provider and persist its state without clobbering the other provider. */
  private async withProvider<T>(p: Provider, fn: (st: ProviderState) => Promise<T>): Promise<T | typeof BUSY> {
    if (this.busy.has(p)) return BUSY;
    this.busy.add(p);
    this.onChange();
    try {
      const st = loadState()[p] ?? {};
      try {
        return await fn(st);
      } finally {
        saveProviderState(p, st);
        this.planResetCheck(p, st.lastSnapshot);
      }
    } finally {
      this.busy.delete(p);
      this.onChange();
    }
  }

  /**
   * Besides the regular interval, check once right after the running session's reset
   * time, so the next session starts within seconds instead of up to an interval later.
   * Only while the scheduler is running, so one-off CLI commands exit normally.
   */
  private planResetCheck(p: Provider, snap: Snapshot | undefined) {
    if (!this.timer) return;
    clearTimeout(this.resetTimers.get(p));
    this.resetTimers.delete(p);
    const w = snap?.fiveHour;
    if (!w?.active || w.resetsAt === null) {
      this.resetRetries.delete(p);
      return;
    }
    let delay = w.resetsAt + timing.resetGraceMs[p] - Date.now();
    if (delay <= 0) {
      // Past its reset time but still reported as running: ask again shortly, a few
      // times, then leave it to the regular interval.
      const n = (this.resetRetries.get(p) ?? 0) + 1;
      if (n > MAX_RESET_RETRIES) return;
      this.resetRetries.set(p, n);
      delay = timing.resetRetryMs;
    } else {
      this.resetRetries.delete(p);
    }
    this.resetTimers.set(
      p,
      setTimeout(() => void this.checkAtReset(p), delay),
    );
  }

  private async checkAtReset(p: Provider) {
    this.resetTimers.delete(p);
    const cfg = loadConfig();
    // Outside active hours the first regular check once they begin takes over.
    if (!cfg[p].enabled || !inActiveHours(cfg, Date.now())) return;
    log({ provider: p, event: "reset-check", retry: this.resetRetries.get(p) ?? 0 });
    // If a check is already running, its own completion plans the next reset check.
    await this.withProvider(p, (st) => tickProvider(p, cfg, st, Date.now(), (s) => saveProviderState(p, s)));
  }
}

File: src/go/core/scheduler.go
package core

import (
	"fmt"
	"strconv"
	"strings"
	"sync"
	"time"
)

// startCooldownMs: don't send another start message within this long of the last one.
const startCooldownMs = 30 * 60_000
const failureNotifyAfter = 3
const maxResetRetries = 5

// Timing is overridable by tests.
var Timing = struct {
	// ConfirmDelay is the wait before confirming a start: right after a start, Codex's
	// new window is indistinguishable from its "no window" report (see IdleToleranceMs).
	ConfirmDelay time.Duration
	// ResetGrace is how long after a reported reset time to check again. Codex reports
	// exact seconds; Claude's /usage shows minutes only ("4:59pm" for a 17:00 reset),
	// so it gets more room: checking early would send the start message into the old
	// session.
	ResetGrace map[Provider]time.Duration
	// ResetRetry spaces retries when a session is still reported as running after its
	// reset time.
	ResetRetry time.Duration
}{
	ConfirmDelay: 90 * time.Second,
	ResetGrace:   map[Provider]time.Duration{Claude: 90 * time.Second, Codex: 30 * time.Second},
	ResetRetry:   60 * time.Second,
}

// Check and Start are swappable by tests.
var (
	Check = map[Provider]func(Config) (Snapshot, error){Claude: CheckClaude, Codex: CheckCodex}
	Start = map[Provider]func(Config) (string, error){Claude: StartClaude, Codex: StartCodex}
)

func minutesOfDay(hhmm string) int {
	h, m, _ := strings.Cut(hhmm, ":")
	hi, _ := strconv.Atoi(h)
	mi, _ := strconv.Atoi(m)
	return hi*60 + mi
}

// InActiveHours supports ranges past midnight; nil means always.
func InActiveHours(ah *ActiveHours, now int64) bool {
	if ah == nil {
		return true
	}
	t := time.UnixMilli(now)
	cur := t.Hour()*60 + t.Minute()
	s, e := minutesOfDay(ah.Start), minutesOfDay(ah.End)
	if s <= e {
		return cur >= s && cur < e
	}
	return cur >= s || cur < e
}

// ActiveHoursBegan returns when the active-hours period containing now began, or
// false when now is outside it (or there are no active hours).
func ActiveHoursBegan(ah *ActiveHours, now int64) (int64, bool) {
	if ah == nil || !InActiveHours(ah, now) {
		return 0, false
	}
	s := minutesOfDay(ah.Start)
	t := time.UnixMilli(now)
	d := time.Date(t.Year(), t.Month(), t.Day(), s/60, s%60, 0, 0, time.Local)
	if d.UnixMilli() > now {
		d = d.AddDate(0, 0, -1) // began yesterday (range past midnight)
	}
	return d.UnixMilli(), true
}

type Decision struct {
	Start  bool
	Reason string
}

func ShouldStart(snap Snapshot, st *ProviderState, now int64, cfg Config) Decision {
	switch {
	case snap.FiveHour.Active:
		return Decision{false, "5h session active"}
	case !cfg.AutoStart:
		return Decision{false, "auto-start paused"}
	case snap.Weekly != nil && snap.Weekly.UsedPct >= 100:
		return Decision{false, "weekly limit exhausted"}
	case !InActiveHours(cfg.ActiveHours, now):
		return Decision{false, "outside active hours"}
	case st.LastStartAt != 0 && now-st.LastStartAt < startCooldownMs:
		return Decision{false, "started recently"}
	}
	return Decision{true, "5h session idle"}
}

// SessionActive reports whether a snapshot shows a 5h session that's still running
// (unknown counts as not).
func SessionActive(snap *Snapshot, now int64) bool {
	if snap == nil || !snap.FiveHour.Active {
		return false
	}
	return snap.FiveHour.ResetsAt == nil || *snap.FiveHour.ResetsAt > now
}

func nowMs() int64 { return time.Now().UnixMilli() }

// TickProvider checks one provider and, if its session is idle, starts a new one.
// Mutates st. persist is called as soon as a start succeeds, so the cooldown survives
// a quit or crash during the confirmation delay. starting, if set, is called when it
// sends the start message; the start and its confirmation take a couple of minutes.
func TickProvider(p Provider, cfg Config, st *ProviderState, now int64, persist func(*ProviderState), starting func()) {
	st.LastCheckAt = now
	phase := "check"
	err := func() error {
		snap, err := Check[p](cfg)
		if err != nil {
			return err
		}
		st.LastSnapshot = &snap
		d := ShouldStart(snap, st, now, cfg)
		st.LastDecision = d.Reason
		Log("provider", p, "event", "check", "fiveHour", snap.FiveHour, "weekly", snap.Weekly, "decision", d.Reason)
		if !d.Start {
			return nil
		}

		phase = "start"
		if starting != nil {
			starting()
		}
		reply, err := Start[p](cfg)
		if err != nil {
			return err
		}
		st.LastStartAt = nowMs()
		if persist != nil {
			persist(st)
		}
		phase = "check"
		time.Sleep(Timing.ConfirmDelay)
		after, err := Check[p](cfg)
		if err != nil {
			return err
		}
		st.LastSnapshot = &after
		st.LastDecision = "start sent, session not reported yet"
		msg := fmt.Sprintf("%s start message sent, but no active session reported yet", Label[p])
		if after.FiveHour.Active {
			st.LastDecision = "5h session active"
			msg = fmt.Sprintf("%s 5h session started — resets %s", Label[p], FmtTime(after.FiveHour.ResetsAt))
		}
		Log("provider", p, "event", "start", "model", cfg.Model(p), "reply", reply, "fiveHour", after.FiveHour)
		Notify("Usage Window Starter", msg)
		return nil
	}()
	if err == nil {
		st.ConsecutiveFailures = 0
		st.FailureNotified = false
		st.LastError = ""
		return
	}
	msg := err.Error()
	st.ConsecutiveFailures++
	st.LastError = msg
	Log("provider", p, "event", "error", "phase", phase, "error", msg, "consecutive", st.ConsecutiveFailures)
	// A failed start (e.g. a model that's no longer offered) needs attention right away.
	if (phase == "start" || st.ConsecutiveFailures >= failureNotifyAfter) && !st.FailureNotified {
		st.FailureNotified = true
		if phase == "start" {
			Notify("Usage Window Starter", fmt.Sprintf(`%s couldn't start a 5h session with model "%s": %s`, Label[p], cfg.Model(p), truncate(msg, 100)))
		} else {
			Notify("Usage Window Starter", fmt.Sprintf("%s limit check failing: %s", Label[p], truncate(msg, 120)))
		}
	}
}

// Scheduler runs checks on the configured interval. It wakes every minute and
// compares wall-clock time, so checks catch up right after the Mac wakes from sleep.
type Scheduler struct {
	OnChange func()

	mu sync.Mutex
	// activity is "checking" or "starting" while a provider is busy.
	activity     map[Provider]string
	stop         chan struct{}
	resetTimers  map[Provider]*time.Timer
	resetRetries map[Provider]int
}

func NewScheduler(onChange func()) *Scheduler {
	if onChange == nil {
		onChange = func() {}
	}
	return &Scheduler{
		OnChange:     onChange,
		activity:     map[Provider]string{},
		resetTimers:  map[Provider]*time.Timer{},
		resetRetries: map[Provider]int{},
	}
}

func (s *Scheduler) Run() {
	s.mu.Lock()
	if s.stop != nil {
		s.mu.Unlock()
		return
	}
	stop := make(chan struct{})
	s.stop = stop
	s.mu.Unlock()

	go func() {
		t := time.NewTicker(time.Minute)
		defer t.Stop()
		for {
			select {
			case <-stop:
				return
			case <-t.C:
				s.RunDue(false, nowMs())
			}
		}
	}()
	// The first RunDue may skip providers checked recently (e.g. right after a
	// restart), so re-arm reset checks from the saved snapshots.
	cfg := LoadConfig()
	state := LoadState()
	for _, p := range Providers {
		if cfg.Enabled(p) && state[p] != nil {
			s.planResetCheck(p, state[p].LastSnapshot)
		}
	}
	go s.RunDue(false, nowMs())
}

func (s *Scheduler) Stop() {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.stop != nil {
		close(s.stop)
		s.stop = nil
	}
	for p, t := range s.resetTimers {
		t.Stop()
		delete(s.resetTimers, p)
	}
}

func (s *Scheduler) IsBusy(p Provider) bool { return s.Activity(p) != "" }

// Activity is what provider p is doing: "checking", "starting", or "" when idle.
func (s *Scheduler) Activity(p Provider) string {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.activity[p]
}

func (s *Scheduler) setActivity(p Provider, a string) {
	s.mu.Lock()
	s.activity[p] = a
	s.mu.Unlock()
	s.OnChange()
}

// tick runs TickProvider with state persistence and activity reporting.
func (s *Scheduler) tick(p Provider, cfg Config, st *ProviderState, now int64) {
	TickProvider(p, cfg, st, now, func(st *ProviderState) { SaveProviderState(p, st) },
		func() { s.setActivity(p, "starting") })
}

func (s *Scheduler) running() bool {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.stop != nil
}

// RunDue checks every enabled provider whose interval has elapsed (or all of them
// with force). Nothing runs outside active hours unless forced. When they begin,
// every provider not checked since is due at once, instead of each waiting out its
// own interval.
func (s *Scheduler) RunDue(force bool, now int64) {
	cfg := LoadConfig()
	if !force && !InActiveHours(cfg.ActiveHours, now) {
		return
	}
	began, inHours := ActiveHoursBegan(cfg.ActiveHours, now)
	s.runWhere(cfg, now, func(last int64) bool {
		return force || now-last >= int64(cfg.IntervalMin)*60_000 || (inHours && last < began)
	})
}

// Refresh checks every enabled provider not checked within maxAge, in active hours only.
func (s *Scheduler) Refresh(maxAge time.Duration, now int64) {
	cfg := LoadConfig()
	if !InActiveHours(cfg.ActiveHours, now) {
		return
	}
	s.runWhere(cfg, now, func(last int64) bool { return now-last >= maxAge.Milliseconds() })
}

func (s *Scheduler) runWhere(cfg Config, now int64, isDue func(lastCheckAt int64) bool) {
	state := LoadState()
	var wg sync.WaitGroup
	for _, p := range Providers {
		var last int64
		if state[p] != nil {
			last = state[p].LastCheckAt
		}
		if !cfg.Enabled(p) || s.IsBusy(p) || !isDue(last) {
			continue
		}
		wg.Add(1)
		go func(p Provider) {
			defer wg.Done()
			s.withProvider(p, func(st *ProviderState) {
				s.tick(p, cfg, st, now)
			})
		}(p)
	}
	wg.Wait()
}

// StartNow starts a session right now if it's idle, ignoring pause, cooldown and
// active hours.
func (s *Scheduler) StartNow(p Provider) {
	cfg := LoadConfig()
	cfg.AutoStart = true
	cfg.ActiveHours = nil
	s.withProvider(p, func(st *ProviderState) {
		st.LastStartAt = 0
		st.FailureNotified = false
		s.tick(p, cfg, st, nowMs())
	})
}

// TestModel sends one test message with an unsaved config. It shares the provider
// lock with checks and records a start, because a test message starts a session when
// none is running.
func (s *Scheduler) TestModel(p Provider, cfg Config) (string, error) {
	var reply string
	var err error
	ok := s.withProvider(p, func(st *ProviderState) {
		// Only a session that looked idle can have been started by the test. Recording
		// it during an active session would push the next auto-start back by the cooldown.
		wasIdle := !SessionActive(st.LastSnapshot, nowMs())
		reply, err = Start[p](cfg)
		if err != nil {
			return
		}
		if wasIdle {
			st.LastStartAt = nowMs()
		}
		Log("provider", p, "event", "test", "model", cfg.Model(p), "reply", reply, "countedAsStart", wasIdle)
	})
	if !ok {
		doing := "a check"
		if s.Activity(p) == "starting" {
			doing = "starting a session"
		}
		return "", fmt.Errorf("%s is busy with %s, try again in a moment", Label[p], doing)
	}
	return reply, err
}

// withProvider serializes work per provider and persists its state without
// clobbering the other provider. Returns false when p is already busy.
func (s *Scheduler) withProvider(p Provider, fn func(st *ProviderState)) bool {
	s.mu.Lock()
	if s.activity[p] != "" {
		s.mu.Unlock()
		return false
	}
	s.activity[p] = "checking"
	s.mu.Unlock()
	s.OnChange()

	st := LoadState()[p]
	if st == nil {
		st = &ProviderState{}
	}
	defer func() {
		SaveProviderState(p, st)
		s.planResetCheck(p, st.LastSnapshot)
		s.mu.Lock()
		delete(s.activity, p)
		s.mu.Unlock()
		s.OnChange()
	}()
	fn(st)
	return true
}

// planResetCheck: besides the regular interval, check once right after the running
// session's reset time, so the next session starts within seconds instead of up to
// an interval later. Only while the scheduler is running, so one-off CLI commands
// exit normally.
func (s *Scheduler) planResetCheck(p Provider, snap *Snapshot) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.stop == nil {
		return
	}
	if t := s.resetTimers[p]; t != nil {
		t.Stop()
		delete(s.resetTimers, p)
	}
	if snap == nil || !snap.FiveHour.Active || snap.FiveHour.ResetsAt == nil {
		delete(s.resetRetries, p)
		return
	}
	delay := time.Duration(*snap.FiveHour.ResetsAt-nowMs())*time.Millisecond + Timing.ResetGrace[p]
	if delay <= 0 {
		// Past its reset time but still reported as running: ask again shortly, a few
		// times, then leave it to the regular interval.
		n := s.resetRetries[p] + 1
		if n > maxResetRetries {
			return
		}
		s.resetRetries[p] = n
		delay = Timing.ResetRetry
	} else {
		delete(s.resetRetries, p)
	}
	s.resetTimers[p] = time.AfterFunc(delay, func() { s.checkAtReset(p) })
}

func (s *Scheduler) checkAtReset(p Provider) {
	s.mu.Lock()
	delete(s.resetTimers, p)
	retry := s.resetRetries[p]
	s.mu.Unlock()
	cfg := LoadConfig()
	// Outside active hours the first regular check once they begin takes over.
	if !cfg.Enabled(p) || !InActiveHours(cfg.ActiveHours, nowMs()) {
		return
	}
	Log("provider", p, "event", "reset-check", "retry", retry)
	// If a check is already running, its own completion plans the next reset check.
	s.withProvider(p, func(st *ProviderState) {
		s.tick(p, cfg, st, nowMs())
	})
}
